javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,385 @@
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(next: Node[E], e: E): Unit = {
+    if (next eq head) addFirst(e)
+    else if (next eq null) addLast(e)
+    else {
+      var current = next
+
+      val node = new Node(e, prev = current.prev, next = current)
+      current.prev.next = node
+      current.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    if (index == 0) addFirst(element)
+    else if (index == size) addLast(element)
+    else addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) removeFirst()
+    else if (node eq last) removeLast()
+    else {
+      var current = node
+
+      current.prev.next = current.next
+      current.next.prev = current.prev
+
+      _size -= 1
+
+      current.value
+    }
+  }
+
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    if (index == 0) removeFirst()
+    else if (index == (size - 1)) removeLast()
+    else removeNode(getNodeAt(index))
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    removeFirst()
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E =
+    head.value
+
+  def peekLast(): E =
+    if (last ne null) last.value
+    else peekFirst()
+
+  def pollFirst(): E =
+    removeFirst()
+
+  def pollLast(): E =
+    removeLast()
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  private def _removeOccurrence(iter: Iterator[E], o: Any): Boolean = {
+    var changed = false
+    val e = o.asInstanceOf[E]
+    while (iter.hasNext() && !changed) {
+      if (iter.next() === e) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean =
+    _removeOccurrence(iterator(), o)
+
+  def removeLastOccurrence(o: Any): Boolean =
+    _removeOccurrence(descendingIterator(), o)
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new ListIterator[E] {
+
+      private var last = -1
+      private var i = index
+
+      private var currentNode: Node[E] =
+        if (index == size) null
+        else getNodeAt(index)
+
+      private var lastNode: Node[E] =
+        if (currentNode eq null) getNodeAt(size - 1)
+        else null
```

Comment:
Could you explain what the invariants regarding `lastNode`? What does it represent?

I would have assumed that it was always the node at index `last` (or `null` if `last == -1`), i.e. that it is the node that would be affected by `remove()` or `set()`. But if that is the case, it should always be initialized to `null` (since `last` is initialized to -1).